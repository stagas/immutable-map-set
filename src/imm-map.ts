export class ImmMap<K, V> extends Map<K, V> {
  constructor(entries?: readonly (readonly [K, V])[] | Map<K, V> | ImmMap<K, V> | null) {
    super()
    if (entries)
      for (const [key, value] of entries)
        this.mutateSet(key, value)
  }
  toJSON() {
    return [...this]
  }
  // @ts-ignore
  set(key: K, value: V): ImmMap<K, V> {
    return new ImmMap(this).mutateSet(key, value)
  }
  mutateSet(key: K, value: V): this {
    return super.set(key, value)
  }
  // @ts-ignore
  delete(key: K): ImmMap<K, V> {
    const map = new ImmMap(this)
    map.mutateDelete(key)
    return map
  }
  mutateDelete(key: K): boolean {
    return super.delete(key)
  }
  clear(): ImmMap<K, V> {
    return new ImmMap()
  }
}
