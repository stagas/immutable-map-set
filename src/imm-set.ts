export class ImmSet<T> extends Set<T> {
  constructor(values?: readonly T[] | Set<T> | ImmSet<T> | null) {
    super()
    if (values)
      for (const value of values)
        this.mutateAdd(value)
  }
  toJSON() {
    return [...this]
  }
  // @ts-ignore
  add(value: T): ImmSet<T> {
    return new ImmSet(this).mutateAdd(value)
  }
  mutateAdd(value: T): this {
    return super.add(value)
  }
  // @ts-ignore
  delete(value: T): ImmSet<T> {
    const set = new ImmSet(this)
    set.mutateDelete(value)
    return set
  }
  mutateDelete(value: T): boolean {
    return super.delete(value)
  }
  clear(): ImmSet<T> {
    return new ImmSet()
  }
}
