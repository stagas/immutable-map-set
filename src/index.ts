export * from './imm-map'
export * from './imm-set'
