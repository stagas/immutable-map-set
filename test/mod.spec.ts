import { ImmMap, ImmSet } from '../src'

describe('suite', () => {
  describe('ImmMap', () => {
    it('works', () => {
      const map = new ImmMap([[1, 2]])
      const res = map.set(3, 4)
      expect(map.has(1)).toBe(true)
      expect(map.has(3)).toBe(false)
      expect(res.has(3)).toBe(true)
      expect(res).toBeInstanceOf(ImmMap)
      expect(map).not.toBe(res)
      expect(map).toMatchSnapshot()
      expect(res).toMatchSnapshot()
      const del = res.delete(1)
      expect(del).toBeInstanceOf(ImmMap)
      expect(res).toMatchSnapshot()
      expect(del).toMatchSnapshot()
      expect(del).not.toBe(res)
      const cleared = del.clear()
      expect(cleared).not.toBe(del)
      expect(del).toMatchSnapshot()
      expect(cleared).toBeInstanceOf(ImmMap)
      expect(cleared).toMatchSnapshot()
    })
  })

  describe('ImmSet', () => {
    it('works', () => {
      const set = new ImmSet([1, 2])
      const res = set.add(3)
      expect(set.has(1)).toBe(true)
      expect(set.has(3)).toBe(false)
      expect(res.has(3)).toBe(true)
      expect(res).toBeInstanceOf(ImmSet)
      expect(set).not.toBe(res)
      expect(set).toMatchSnapshot()
      expect(res).toMatchSnapshot()
      const del = res.delete(1)
      expect(del).toBeInstanceOf(ImmSet)
      expect(res).toMatchSnapshot()
      expect(del).toMatchSnapshot()
      expect(del).not.toBe(res)
      const cleared = del.clear()
      expect(cleared).not.toBe(del)
      expect(del).toMatchSnapshot()
      expect(cleared).toBeInstanceOf(ImmSet)
      expect(cleared).toMatchSnapshot()
    })
  })
})
