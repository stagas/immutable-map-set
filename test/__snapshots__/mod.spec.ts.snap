// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suite ImmMap works 1`] = `
Map {
  1 => 2,
}
`;

exports[`suite ImmMap works 2`] = `
Map {
  1 => 2,
  3 => 4,
}
`;

exports[`suite ImmMap works 3`] = `
Map {
  1 => 2,
  3 => 4,
}
`;

exports[`suite ImmMap works 4`] = `
Map {
  3 => 4,
}
`;

exports[`suite ImmMap works 5`] = `
Map {
  3 => 4,
}
`;

exports[`suite ImmMap works 6`] = `Map {}`;

exports[`suite ImmSet works 1`] = `
Set {
  1,
  2,
}
`;

exports[`suite ImmSet works 2`] = `
Set {
  1,
  2,
  3,
}
`;

exports[`suite ImmSet works 3`] = `
Set {
  1,
  2,
  3,
}
`;

exports[`suite ImmSet works 4`] = `
Set {
  2,
  3,
}
`;

exports[`suite ImmSet works 5`] = `
Set {
  2,
  3,
}
`;

exports[`suite ImmSet works 6`] = `Set {}`;
